### PR TITLE
[cpp_info] [refact] Do not assign to 'xxxx_paths' fields

### PR DESCRIPTION
**Pull request by jgsogo**
Changelog: omit
Docs: omit

Attributes like `xxxx_path` in `cpp_info`-related objects are read-only, they cannot be assigned or modified. This change requires other changes too:
 * Tests cannot assign to these fields, assignments are moved to `xxxdirs` attributes, now default values appear!
 * As read-only fields, they could be tuples, lists or generators. Using `list(...)` to get their values abstract this away.
 * It requires `filter_empty=False` so paths are not removed when retrieving them via `xxxx_paths`

### CODE REVIEW

File: conans/test/unittests/model/build_info/components_test.py [hunk 1]
```diff
@@ -50,7 +49,7 @@ def test_deps_cpp_info_libs(self):
         self.assertListEqual(["libc", "libd"], deps_cpp_info["dep2"].libs)
         self.assertListEqual(["libdep3"], deps_cpp_info["dep3"].libs)
         self.assertListEqual(["liba", "libb", "libc", "libd", "libdep3"],
-                             deps_cpp_info.libs)
+                             list(deps_cpp_info.libs))
```
Comment (danimtb) [inline conans/test/unittests/model/build_info/components_test.py:52]: I don't get why this change is needed. Isn't `deps_cpp_info.libs` already a list?

Comment (jgsogo) [reply conans/test/unittests/model/build_info/components_test.py:52]: `libs` attribute from any other object but `cpp_info` itself is a read-only field.

Right now it is a `list`, agree, but that is the worst choice for a read-only field, implementation could change to `tuple` or even a `generator` and the test shouldn't fail because of that _implementation detail_. That's the reason why I'm calling `list`, to expand a possible generator or change a tuple into a list.

**Review by danimtb — APPROVED** [review]

File: conans/test/unittests/model/build_info/components_test.py (hunk 1 above)
Comment (memsharded) [reply conans/test/unittests/model/build_info/components_test.py:52]: Actually, I think this conversion to ``list()`` in the test, that allows changing the implementation of ``.libs``, ``.include_paths`` and other attributes to a different thing other than a list, is a risky thing. Because it is explicitly documented that these things are lists: https://docs.conan.io/en/latest/reference/conanfile/attributes.html#cpp-info

We could have the tests green, but break users without knowing. I think the ``list()`` conversions need to be removed, to guarantee in tests that these attribute return lists.

**Review by memsharded — CHANGES_REQUESTED** [review]
I am concerned about the ``list()`` conversion, as it can hide breaking changes in the model exposed to users.

**Comment by jgsogo** [comment]
`cpp_info.**` will always be a list. The consumer part is `deps_cpp_info`, there it only mentions the _list_ word once, but it says _"In addition of the above fields"_ (which is not totally true, btw)  

I will add a test to check they are a `list` or a `tuple` (**are we ok with a tuple?** 😬 ), to avoid changing them to a generator without noticing.
